cabal-install/src/Distribution/Client/CmdRepl.hs

```diff
@@ -195,19 +190,42 @@ replCommand = Client.installCommand {
 -- "Distribution.Client.ProjectOrchestration"
 --
 replAction :: NixStyleFlags (ReplOptions, EnvFlags) -> [String] -> GlobalFlags -> IO ()
-replAction flags@NixStyleFlags { extraFlags = (replFlags, envFlags), ..} targetStrings globalFlags = do
-    let
-      with                 = withProject    cliConfig                   verbosity targetStrings
-      without globalConfig = withoutProject (globalConfig <> cliConfig) verbosity targetStrings
-
-    (baseCtx, targetSelectors, finalizer, replType) <-
-      withProjectOrGlobalConfig verbosity ignoreProject globalConfigFlag with without
-
-    when (buildSettingOnlyDeps (buildSettings baseCtx)) $
+replAction flags@NixStyleFlags { extraFlags = (replFlags, envFlags), ..} targetStrings globalFlags
+  = withContextAndSelectors AcceptNoTargets (Just LibKind) flags targetStrings globalFlags $ \targetCtx ctx targetSelectors -> do
+    when (buildSettingOnlyDeps (buildSettings ctx)) $
       die' verbosity $ "The repl command does not support '--only-dependencies'. "
           ++ "You may wish to use 'build --only-dependencies' and then "
           ++ "use 'repl'."
 
+    let projectRoot = distProjectRootDirectory $ distDirLayout ctx
+
+    (replType, mScript, baseCtx) <- case targetCtx of
+      ProjectContext -> return (ProjectRepl, Nothing, ctx)
+      GlobalContext  -> do
+        unless (null targetStrings) $
+          die' verbosity $ "'repl' takes no arguments or a script argument outside a project: " ++ unwords targetStrings
+
+        let
+          sourcePackage = fakeProjectSourcePackage projectRoot
+            & lSrcpkgDescription . L.condLibrary
+            .~ Just (CondNode library [baseDep] [])
+          library = emptyLibrary { libBuildInfo = lBuildInfo }
+          lBuildInfo = emptyBuildInfo
+            { targetBuildDepends = [baseDep]
+            , defaultLanguage = Just Haskell2010
+            }
+          baseDep = Dependency "base" anyVersion mainLibSet
+
+        (,,) GlobalRepl Nothing <$> updateContextAndWriteProjectFile' ctx sourcePackage
```

Comment:
Gotta be honest, I hate this :D

TupleSections maybe? 
```suggestion
        (GlobalRepl,Nothing,) <$> updateContextAndWriteProjectFile' ctx sourcePackage
```

If you prefer it this way, your choice!